libraries/Networking/src/ESP_Network_Manager.cpp

```diff
@@ -0,0 +1,136 @@
+#include "ESP_Network_Manager.h"
+#include "esp_netif.h"
+#include "lwip/ip_addr.h"
+#include "lwip/dns.h"
+#include "esp32-hal-log.h"
+#include "esp_mac.h"
+
+ESP_Network_Manager::ESP_Network_Manager(){
+
+}
+
+bool ESP_Network_Manager::begin(){
+    static bool initialized = false;
+    if(!initialized){
+        initialized = true;
+#if CONFIG_IDF_TARGET_ESP32
+        uint8_t mac[8];
+        if(esp_efuse_mac_get_default(mac) == ESP_OK){
+            esp_base_mac_addr_set(mac);
+        }
+#endif
+        initialized = esp_netif_init() == ESP_OK;
+        if(!initialized){
+        	log_e("esp_netif_init failed!");
+        }
+    }
+    if(initialized){
+    	initNetworkEvents();
+    }
+    return initialized;
+}
+
+typedef struct gethostbynameParameters {
+    const char *hostname;
+    ip_addr_t addr;
+    uint8_t addr_type;
+    int result;
+} gethostbynameParameters_t;
+
+/**
+ * DNS callback
+ * @param name
+ * @param ipaddr
+ * @param callback_arg
+ */
+static void wifi_dns_found_callback(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
+{
+    gethostbynameParameters_t *parameters = static_cast<gethostbynameParameters_t *>(callback_arg);
+    if(ipaddr) {
+        if(parameters->result == 0){
+            memcpy(&(parameters->addr), ipaddr, sizeof(ip_addr_t));
+            parameters->result = 1;
+        }
+    } else {
+        parameters->result = -1;
+    }
+    Network.setStatusBits(NET_DNS_DONE_BIT);
+}
+
+/**
+ * Callback to execute dns_gethostbyname in lwIP's TCP/IP context
+ * @param param Parameters for dns_gethostbyname call
+ */
+static esp_err_t wifi_gethostbyname_tcpip_ctx(void *param)
+{
+    gethostbynameParameters_t *parameters = static_cast<gethostbynameParameters_t *>(param);
+    return dns_gethostbyname_addrtype(parameters->hostname, &parameters->addr, &wifi_dns_found_callback, parameters, parameters->addr_type);
+}
+
+/**
+ * Resolve the given hostname to an IP address.
+ * @param aHostname     Name to be resolved
+ * @param aResult       IPAddress structure to store the returned IP address
+ * @return 1 if aIPAddrString was successfully converted to an IP address,
+ *          else error code
+ */
+int ESP_Network_Manager::hostByName(const char* aHostname, IPAddress& aResult, bool preferV6)
+{
+    err_t err = ERR_OK;
+    gethostbynameParameters_t params;
+
+    // This should generally check if we have a global address assigned to one of the interfaces.
+    // If such address is not assigned, there is no point in trying to get V6 from DNS as we will not be able to reach it.
+    // That is of course, if 'preferV6' is not set to true
+    static bool hasGlobalV6 = false;
+    bool hasGlobalV6Now = (Network.getStatusBits() & NET_HAS_IP6_GLOBAL_BIT) != 0;
```

Comment:
ULA addresses should probably also support IPv6 lookup in DNS.

They may be used in private deployments, with either direct DNS support, or NAT66 (like NAT44) where they can connect to public addresses

Reply:
For example my Thread network at home (Google) uses random ULA ranges to addresses; an Eve switch I own has the addresses:
FE 80 00 00 00 00 00 00 C0 C8 06 A1 78 37 2F 37
FD 62 CB DC AC 0A 00 00 D5 0C F4 56 0A 5B 02 27
FD 1B 49 A9 44 D5 00 01 FA 37 8B B6 11 70 35 6F

I haven't tested a device on it yet, but for this device to reach a public address would use NAT66.  I have seen mention of OTBR (Open Thread Border Router, the underlying open source) mentions NAT64.

In either case DNS (or in the case of Thread mDNS) would need to return IPv6 addresses to be of use.

So, we should do an IPv6 lookup for either ESP GLOBAL   or ULA.